gpytorch/lazy/lazy_tensor.py

```diff
@@ -732,6 +741,218 @@ def add_jitter(self, jitter_val=1e-3):
         diag = torch.tensor(jitter_val, dtype=self.dtype, device=self.device)
         return self.add_diag(diag)
 
+    def cat_rows(self, cross_mat, new_mat, generate_roots=True, **root_decomp_kwargs):
+        """
+        Concatenates new rows and columns to the matrix that this LazyTensor represents, e.g.
+        C = ((A; B); B^T, D). where A is the existing lazy tensor, and B (cross_mat) and D (new_mat)
+        are new components. This is most commonly used when fantasizing with kernel matrices.
+
+        We have access to K \\approx LL^T and K^{-1} \\approx RR^T, where L and R are low rank matrices
+        resulting from root and root inverse decompositions (see the LOVE paper).
+
+        To update R, we first update L:
+            [K U; U^T S] = [L 0; A B][L^T A^T; 0 BT]
+        Solving this matrix equation, we get:
+            K = LL^T ==>       L = L
+            U = LA^T ==>       A = UR^{-1}
+            S = AA^T + BB^T ==> B = cholesky(S - AA^T)
+
+        Once we've computed Z = [L 0; A B], we have that the new kernel matrix [K U; U^T S] \approx ZZ^T. Therefore,
+        we can form a pseudo-inverse of Z directly to approximate [K U; U^T S]^{-1/2}.
+
+        This strategy is also described in "Efficient Nonmyopic Bayesian Optimization via One-Shot Multistep Trees,"
+        Jiang et al, NeurIPS, 2020. https://arxiv.org/abs/2006.15779.
+
+        Args:
+            cross_mat (:obj:`torch.tensor`): the matrix :math:`B` we are appending to the matrix :math:`A`.
+                If :math:`A` is n x n, then this matrix should be n x k.
+            new_mat (:obj:`torch.tensor`): the matrix :math:`D` we are appending to the matrix :math:`A`.
+                If :math:`B` is n x k, then this matrix should be k x k.
+            generate_roots (:obj:`bool`): whether to generate the root decomposition of :math:`A` even if it
+                has not been created yet.
+
+        Returns:
+            :obj:`LazyTensor`: concatenated lazy tensor with the new rows and columns.
+        """
+        from . import lazify
+        from .batch_repeat_lazy_tensor import BatchRepeatLazyTensor
+        from .cat_lazy_tensor import CatLazyTensor
+        from .root_lazy_tensor import RootLazyTensor
+        from .triangular_lazy_tensor import TriangularLazyTensor
+
+        cross_mat = lazify(cross_mat)
+        batch_shape = cross_mat.shape[:-2]
+        new_mat = lazify(new_mat)
+        old_tsr = self
+        if old_tsr.ndimension() < cross_mat.ndimension():
+            old_tsr = BatchRepeatLazyTensor(old_tsr, batch_shape)
```

Comment:
@wjmaddox I believe this breaks the broadcasting that we do when computing fantasy models. The issue is that `old_tsr` can itself be batched, so that the `BatchRepeatLazyTensor` messes up the dimensions here. 

Specifically, say `old_tsr` is `3 x 4 x 4` and `cross_mat` is of shape `2 x 3 x 2 x 4` then this `BatchRepeatLazyTensor` will be of shape `2 x 9 x 4 x 4`. Let me try to fix this consistently